core/crypto/rng_hw.c

```diff
@@ -3,19 +3,89 @@
 
 #include <compiler.h>
 #include <crypto/crypto.h>
+#include <kernel/mutex.h>
 #include <kernel/panic.h>
+#include <malloc.h>
 #include <rng_support.h>
 #include <tee/tee_cryp_utl.h>
 #include <types_ext.h>
 
-/* This is a HW RNG, no need for seeding */
+static uint8_t *rng_fifo;
+static size_t rng_fifo_size;
+static size_t rng_fifo_pos;
+static struct mutex rng_fifo_mutex = MUTEX_INITIALIZER;
+
+TEE_Result __weak hw_get_max_available_entropy(size_t *blen)
+{
+	*blen = sizeof(size_t);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result __weak hw_get_available_entropy(void *buf)
+{
+	TEE_Result ret;
+
+	ret = hw_get_random_bytes(buf, sizeof(size_t));
```

Comment:
this call may block. Getting available entropy is not expected to block. I think the sole weak implementation of `hw_get_available_entropy()` is to return `TEE_ERROR_NOT_SUPPORTED`.

If driver provide an implementation for `hw_get_available_entropy()`, then fine. If not, there ain't any.

We can still implement a weak generic `hw_get_random_bytes()` that calls `hw_get_available_entropy()`. No need for mutex in that function.

Here the question is does platform driver `hw_get_available_entropy()` caches generated random bytes or is it done from a generic function, in _rng_hw.c_.

Reply:
Sounds fair, lets take Jens' suggestion and end this PR before we add this `hw_get_available_entropy()` and take the last two patches of this series into their own. We can continue there.